lms/util/h.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+"""Utility functions for working with the h API."""
+
+from __future__ import unicode_literals
+
+import random
+import string
+
+
+ALLOWED_CHARS = string.ascii_letters + string.digits + "._"
+"""The characters that are allowed in h usernames."""
+
+
+def generate_username(request_params):
+    """Return an h username generated from the given LTI launch request params."""
+
+    full_name = request_params.get("lis_person_name_full") or ""
+    user_id = request_params.get("user_id") or ""
+
+    # Take up to 22 valid characters from full_name.
+    # (h usernames can be up to 30 chars long, and we will be appending 8 chars
+    # from user_id below, leaving us up to 22 chars from full_name).
+    sanitized_full_name = _sanitize_string(full_name, limit=22) or "Anonymous"
+
+    sanitized_user_id = _pad_with_random_characters(
+        _sanitize_string(user_id, limit=8), 8
+    )
+
+    return sanitized_full_name + sanitized_user_id
+
+
+def _sanitize_string(unsanitized_string, limit):
+    """Return ``unsanitized_string`` with invalid chars removed.
+
+    Remove any characters not valid in an h username, and also truncate the
+    resulting string of valid characters to ``limit`` characters long.
+    """
+    sanitized_string = ""
+    for char in unsanitized_string:
+        if char in ALLOWED_CHARS:
+            sanitized_string = sanitized_string + char
```

Comment:
Here (and below too) we're generating strings inefficiently by appending one character at a time. I think this is fine since we're dealing with 30-character strings here.